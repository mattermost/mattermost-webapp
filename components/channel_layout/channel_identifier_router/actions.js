// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {joinChannel, getChannelByNameAndTeamName, markGroupChannelOpen, fetchMyChannelsAndMembers} from 'mattermost-redux/actions/channels';
import {getUser, getUserByUsername, getUserByEmail} from 'mattermost-redux/actions/users';
import {getTeamByName} from 'mattermost-redux/selectors/entities/teams';
import {getCurrentUserId, getUserByUsername as selectUserByUsername, getUser as selectUser, getUserByEmail as selectUserByEmail} from 'mattermost-redux/selectors/entities/users';
import {getChannelByName, getOtherChannels, getChannel, getChannelsNameMapInTeam, getRedirectChannelNameForTeam} from 'mattermost-redux/selectors/entities/channels';

import {Constants} from 'utils/constants';
import {openDirectChannelToUserId} from 'actions/channel_actions';
import * as GlobalActions from 'actions/global_actions.jsx';
import * as Utils from 'utils/utils.jsx';

const LENGTH_OF_ID = 26;
const LENGTH_OF_GROUP_ID = 40;
const LENGTH_OF_USER_ID_PAIR = 54;
const USER_ID_PAIR_REGEXP = new RegExp(`^[a-zA-Z0-9]{${LENGTH_OF_ID}}__[a-zA-Z0-9]{${LENGTH_OF_ID}}$`);

export function onChannelByIdentifierEnter({match, history}) {
    return async (dispatch, getState) => {
        const state = getState();
        const {path, identifier, team} = match.params;

        if (!identifier) {
            return;
        }

        const teamObj = getTeamByName(state, team);
        if (!teamObj) {
            return;
        }

        const channelPath = getPathFromIdentifier(state, path, identifier);

        switch (channelPath) {
        case 'channel_name':
            dispatch(goToChannelByChannelName(match, history));
            break;
        case 'channel_id':
            dispatch(goToChannelByChannelId(match, history));
            break;
        case 'group_channel_group_id':
            dispatch(goToGroupChannelByGroupId(match, history));
            break;
        case 'direct_channel_username':
            dispatch(goToDirectChannelByUsername(match, history));
            break;
        case 'direct_channel_email':
            dispatch(goToDirectChannelByEmail(match, history));
            break;
        case 'direct_channel_user_ids':
            dispatch(goToDirectChannelByUserIds(match, history));
            break;
        case 'direct_channel_user_id':
            dispatch(goToDirectChannelByUserId(match, history, identifier));
            break;
        case 'error':
            await dispatch(fetchMyChannelsAndMembers(teamObj.id));
            handleError(match, history, getRedirectChannelNameForTeam(state, teamObj.id));
            break;
        }
    };
}

export function getPathFromIdentifier(state, path, identifier) {
    if (path === 'channels') {
        // It's hard to tell an ID apart from a channel name of the same length, so check first if
        // the identifier matches a channel that we have
        const channelsByName = getChannelByName(state, identifier);
        const moreChannelsByName = getOtherChannels(state).find((chan) => chan.name === identifier);

        if (identifier.length === LENGTH_OF_ID) {
            return channelsByName || moreChannelsByName ? 'channel_name' : 'channel_id';
        } else if (
            (!channelsByName && !moreChannelsByName && identifier.length === LENGTH_OF_GROUP_ID) ||
            (
                (channelsByName && channelsByName.type === Constants.GM_CHANNEL) ||
                (moreChannelsByName && moreChannelsByName.type === Constants.GM_CHANNEL)
            )
        ) {
            return 'group_channel_group_id';
        } else if (isDirectChannelIdentifier(identifier)) {
            return 'direct_channel_user_ids';
        }
        return 'channel_name';
    } else if (path === 'messages') {
        if (identifier.indexOf('@') === 0) {
            return 'direct_channel_username';
        } else if (identifier.indexOf('@') > 0) {
            return 'direct_channel_email';
        } else if (identifier.length === LENGTH_OF_ID) {
            return 'direct_channel_user_id';
        } else if (identifier.length === LENGTH_OF_GROUP_ID) {
            return 'group_channel_group_id';
        }
        return 'error';
    }

    return 'error';
}

export function goToChannelByChannelId(match, history) {
    return async (dispatch, getState) => {
        const state = getState();
        const {team, identifier} = match.params;
        const channelId = identifier.toLowerCase();

        let channel = getChannel(state, channelId);
        const member = state.entities.channels.myMembers[channelId];
        const teamObj = getTeamByName(state, team);
        if (!channel || !member) {
            const {data, error} = await dispatch(joinChannel(getCurrentUserId(state), teamObj.id, channelId, null));
            if (error) {
                await dispatch(fetchMyChannelsAndMembers(teamObj.id));
                handleChannelJoinError(match, history, getRedirectChannelNameForTeam(state, teamObj.id));
                return;
            }
            channel = data.channel;
        }

        if (channel.type === Constants.DM_CHANNEL) {
            dispatch(goToDirectChannelByUserId(match, history, Utils.getUserIdFromChannelId(channel.name, getCurrentUserId(state))));
        } else if (channel.type === Constants.GM_CHANNEL) {
            history.replace(`/${team}/messages/${channel.name}`);
        } else {
            history.replace(`/${team}/channels/${channel.name}`);
        }
    };
}

export function goToChannelByChannelName(match, history) {
    return async (dispatch, getState) => {
        const state = getState();
        const {team, identifier} = match.params;
        const channelName = identifier.toLowerCase();

        const teamObj = getTeamByName(state, team);
        if (!teamObj) {
            return;
        }

        let channel = getChannelsNameMapInTeam(state, teamObj.id)[channelName];
        let member;
        if (channel) {
            member = state.entities.channels.myMembers[channel.id];
        }

        if (!channel || !member) {
            const {data, error: joinError} = await dispatch(joinChannel(getCurrentUserId(state), teamObj.id, null, channelName));
            if (joinError) {
                const {data: data2, error: getChannelError} = await dispatch(getChannelByNameAndTeamName(team, channelName, true));
                if (getChannelError || data2.delete_at === 0) {
                    await dispatch(fetchMyChannelsAndMembers(teamObj.id));
                    handleChannelJoinError(match, history, getRedirectChannelNameForTeam(state, teamObj.id));
                    return;
                }
                channel = data2;
            } else {
                channel = data.channel;
            }
        }

        if (channel.type === Constants.DM_CHANNEL) {
            dispatch(goToDirectChannelByUserIds(match, history));
        } else if (channel.type === Constants.GM_CHANNEL) {
            history.replace(`/${team}/messages/${channel.name}`);
        } else {
            doChannelChange(channel);
        }
    };
}

function goToDirectChannelByUsername(match, history) {
    return async (dispatch, getState) => {
        const state = getState();
        const {team, identifier} = match.params;
        const username = identifier.slice(1, identifier.length).toLowerCase();
        const teamObj = getTeamByName(state, team);

        let user = selectUserByUsername(state, username);
        if (!user) {
            const {data, error} = await dispatch(getUserByUsername(username));
            if (error) {
                await dispatch(fetchMyChannelsAndMembers(teamObj.id));
                handleError(match, history, getRedirectChannelNameForTeam(state, teamObj.id));
                return;
            }
            user = data;
        }

        const {error, data} = await dispatch(openDirectChannelToUserId(user.id));
        if (error) {
            await dispatch(fetchMyChannelsAndMembers(teamObj.id));
            handleError(match, history, getRedirectChannelNameForTeam(state, teamObj.id));
            return;
        }

        doChannelChange(data);
    };
}

export function goToDirectChannelByUserId(match, history, userId) {
    return async (dispatch, getState) => {
        const state = getState();
        const {team} = match.params;
        const teamObj = getTeamByName(state, team);

        let user = selectUser(state, userId);
        if (!user) {
            const {data, error} = await dispatch(getUser(userId));
            if (error) {
                await dispatch(fetchMyChannelsAndMembers(teamObj.id));
                handleError(match, history, getRedirectChannelNameForTeam(state, teamObj.id));
                return;
            }
            user = data;
        }

        history.replace(`/${team}/messages/@${user.username}`);
    };
}

export function goToDirectChannelByUserIds(match, history) {
    return async (dispatch, getState) => {
        const state = getState();
        const {team, identifier} = match.params;
        const userId = Utils.getUserIdFromChannelId(identifier.toLowerCase(), getCurrentUserId(getState()));
        const teamObj = getTeamByName(state, team);

        let user = selectUser(state, userId);
        if (!user) {
            const {data, error} = await dispatch(getUser(userId));
            if (error) {
                await dispatch(fetchMyChannelsAndMembers(teamObj.id));
                handleError(match, history, getRedirectChannelNameForTeam(state, teamObj.id));
                return;
            }
            user = data;
        }

        history.replace(`/${team}/messages/@${user.username}`);
    };
}

export function goToDirectChannelByEmail(match, history) {
    return async (dispatch, getState) => {
        const state = getState();
        const {team, identifier} = match.params;
        const email = identifier.toLowerCase();
        const teamObj = getTeamByName(state, team);

        let user = selectUserByEmail(state, email);
        if (!user) {
            const {data, error} = await dispatch(getUserByEmail(email));
            if (error) {
                await dispatch(fetchMyChannelsAndMembers(teamObj.id));
                handleError(match, history, getRedirectChannelNameForTeam(state, teamObj.id));
                return;
            }
            user = data;
        }

        history.replace(`/${team}/messages/@${user.username}`);
    };
}

function goToGroupChannelByGroupId(match, history) {
    return async (dispatch, getState) => {
        const state = getState();
        const {identifier, team} = match.params;
        const groupId = identifier.toLowerCase();

        history.replace(match.url.replace('/channels/', '/messages/'));

        let channel = getChannelByName(state, groupId);
        const teamObj = getTeamByName(state, team);
        if (!channel) {
            const {data, error} = await dispatch(joinChannel(getCurrentUserId(state), teamObj.id, null, groupId));
            if (error) {
                await dispatch(fetchMyChannelsAndMembers(teamObj.id));
                handleError(match, history, getRedirectChannelNameForTeam(state, teamObj.id));
                return;
            }
            channel = data.channel;
        }

        dispatch(markGroupChannelOpen(channel.id));

        doChannelChange(channel);
    };
}

function doChannelChange(channel) {
    GlobalActions.emitChannelClickEvent(channel);
}

function handleError(match, history, defaultChannel) {
    const {team} = match.params;
    history.push(team ? `/${team}/channels/${defaultChannel}` : '/');
}

function isDirectChannelIdentifier(identifier) {
    return identifier.length === LENGTH_OF_USER_ID_PAIR && USER_ID_PAIR_REGEXP.test(identifier);
}

async function handleChannelJoinError(match, history, defaultChannel) {
    const {team} = match.params;
    history.push(team ? `/error?type=channel_not_found&returnTo=/${team}/channels/${defaultChannel}` : '/');
}