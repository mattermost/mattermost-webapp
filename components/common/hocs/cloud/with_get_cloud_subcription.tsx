// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React, {ComponentType} from 'react';

import {isEmpty} from 'lodash';

import {Subscription} from 'mattermost-redux/types/cloud';

interface Actions {
    getCloudSubscription: () => void;
}

interface UsedHocProps {
    subscription: Subscription;
    isCloud: boolean;
    actions: Actions;
    userIsAdmin: boolean;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function withGetCloudSubscription<P>(WrappedComponent: ComponentType<P>): ComponentType<any> {
    return class extends React.Component<P & UsedHocProps> {
        async componentDidMount() {
            const {subscription, actions, isCloud, userIsAdmin} = this.props;
            if (isEmpty(subscription) && isCloud && userIsAdmin) {
                await actions.getCloudSubscription();
            }
        }

        render(): JSX.Element {
            return <WrappedComponent {...this.props}/>;
        }
    };
}

export default withGetCloudSubscription;
