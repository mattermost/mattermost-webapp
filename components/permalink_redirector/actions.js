// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {getCurrentUserId} from 'mattermost-redux/selectors/entities/users';
import {getTeam} from 'mattermost-redux/selectors/entities/teams';

import {ErrorPageTypes} from 'utils/constants';
import {browserHistory} from 'utils/browser_history';

import LocalStorageStore from 'stores/local_storage_store';

export const redirect = (url) => {
    return (dispatch, getState) => {
        const sanitizedUrl = url.replace('/_redirect/', '');

        const state = getState();
        const teamId = LocalStorageStore.getPreviousTeamId(getCurrentUserId(state));
        const team = getTeam(state, teamId);

        if (!team) {
            browserHistory.replace(`/error?type=${ErrorPageTypes.TEAM_NOT_FOUND}`);
            return;
        }

        browserHistory.push(`/${team.name}/${sanitizedUrl}`);
    };
};
