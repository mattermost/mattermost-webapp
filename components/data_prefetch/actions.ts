// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {Dispatch} from 'redux';

import {getChannelIdsForCurrentTeam} from 'mattermost-redux/selectors/entities/channels';

import {trackEvent} from 'actions/telemetry_actions';

import {GlobalState} from 'types/store';

let isFirstPreload = true;

export function trackPreloadedChannels(prefetchQueueObj: Record<string, string[]>) {
    return (dispatch: Dispatch, getState: () => GlobalState) => {
        const state = getState();
        const channelIdsForTeam = getChannelIdsForCurrentTeam(state);

        trackEvent('performance', 'preloaded_channels', {
            numHigh: prefetchQueueObj[1]?.length || 0,
            numMedium: prefetchQueueObj[2]?.length || 0,
            numLow: prefetchQueueObj[3]?.length || 0,

            numTotal: channelIdsForTeam.length,

            // Tracks whether this is the first team that we've preloaded channels for in this session since
            // the first preload will likely include DMs and GMs
            isFirstPreload,
        });

        isFirstPreload = false;
    };
}
