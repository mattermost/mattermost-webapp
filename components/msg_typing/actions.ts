// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {GlobalState} from '@mattermost/types/store';

import {getMissingProfilesByIds, getStatusesByIds} from 'mattermost-redux/actions/users';

import {General, Preferences, WebsocketEvents} from 'mattermost-redux/constants';

import {getConfig, isPerformanceDebuggingEnabled} from 'mattermost-redux/selectors/entities/general';
import {getBool} from 'mattermost-redux/selectors/entities/preferences';
import {getCurrentUserId, getStatusForUserId} from 'mattermost-redux/selectors/entities/users';

import {DispatchFunc, GetStateFunc} from 'mattermost-redux/types/actions';

function getTimeBetweenTypingEvents(state: GlobalState) {
    const config = getConfig(state);

    return config.TimeBetweenUserTypingUpdatesMilliseconds === undefined ? 0 : parseInt(config.TimeBetweenUserTypingUpdatesMilliseconds, 10);
}

export function userStartedTyping(userId: string, channelId: string, rootId: string, now: number) {
    return (dispatch: DispatchFunc, getState: GetStateFunc) => {
        const state = getState();

        if (
            isPerformanceDebuggingEnabled(state) &&
            getBool(state, Preferences.CATEGORY_PERFORMANCE_DEBUGGING, Preferences.NAME_DISABLE_TYPING_MESSAGES)
        ) {
            return;
        }

        dispatch({
            type: WebsocketEvents.TYPING,
            data: {
                id: channelId + rootId,
                userId,
                now,
            },
        });

        // Ideally this followup loading would be done by someone else
        dispatch(fillInMissingInfo(userId));

        setTimeout(() => {
            dispatch(userStoppedTyping(userId, channelId, rootId, now));
        }, getTimeBetweenTypingEvents(state));
    };
}

function fillInMissingInfo(userId: string) {
    return async (dispatch: DispatchFunc, getState: GetStateFunc) => {
        const state = getState();
        const currentUserId = getCurrentUserId(state);

        if (userId !== currentUserId) {
            const result = await dispatch(getMissingProfilesByIds([userId]));
            if (result.data && result.data.length > 0) {
                // Already loaded the user status
                return {data: false};
            }
        }

        const status = getStatusForUserId(state, userId);
        if (status !== General.ONLINE) {
            dispatch(getStatusesByIds([userId]));
        }

        return {data: true};
    };
}

export function userStoppedTyping(userId: string, channelId: string, rootId: string, now: number) {
    return {
        type: WebsocketEvents.STOP_TYPING,
        data: {
            id: channelId + rootId,
            userId,
            now,
        },
    };
}
