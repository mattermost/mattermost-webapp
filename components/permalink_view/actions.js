// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {getChannel, selectChannel, joinChannel, getChannelStats} from 'mattermost-redux/actions/channels';
import {getPostThread} from 'mattermost-redux/actions/posts';
import {getCurrentTeamId} from 'mattermost-redux/selectors/entities/teams';
import {getCurrentUserId} from 'mattermost-redux/selectors/entities/users';

import {loadChannelsForCurrentUser} from 'actions/channel_actions.jsx';
import {loadNewDMIfNeeded, loadNewGMIfNeeded} from 'actions/user_actions.jsx';
import {browserHistory} from 'utils/browser_history';
import {ActionTypes, Constants, ErrorPageTypes} from 'utils/constants';

export function focusPost(postId, returnTo = '') {
    return async (dispatch, getState) => {
        const {data} = await dispatch(getPostThread(postId));

        if (!data) {
            browserHistory.replace(`/error?type=${ErrorPageTypes.PERMALINK_NOT_FOUND}&returnTo=${returnTo}`);
            return;
        }

        const state = getState();
        const channelId = data.posts[data.order[0]].channel_id;
        let channel = state.entities.channels.channels[channelId];
        const teamId = getCurrentTeamId(state);

        if (!channel) {
            const {data: channelData} = await dispatch(getChannel(channelId));

            if (!channelData) {
                browserHistory.replace(`/error?type=${ErrorPageTypes.PERMALINK_NOT_FOUND}&returnTo=${returnTo}`);
                return;
            }

            channel = channelData;
        }

        const myMember = state.entities.channels.myMembers[channelId];

        if (!myMember) {
            // If it's a DM or GM channel and we don't have a channel member for it already, user is not a member
            if (channel.type === Constants.DM_CHANNEL || channel.type === Constants.GM_CHANNEL) {
                browserHistory.replace(`/error?type=${ErrorPageTypes.PERMALINK_NOT_FOUND}&returnTo=${returnTo}`);
                return;
            }

            await dispatch(joinChannel(getCurrentUserId(getState()), null, channelId));
        }

        if (channel.team_id && channel.team_id !== teamId) {
            browserHistory.replace(`/error?type=${ErrorPageTypes.PERMALINK_NOT_FOUND}&returnTo=${returnTo}`);
            return;
        }

        if (channel && channel.type === Constants.DM_CHANNEL) {
            loadNewDMIfNeeded(channel.id);
        } else if (channel && channel.type === Constants.GM_CHANNEL) {
            loadNewGMIfNeeded(channel.id);
        }

        dispatch(selectChannel(channelId));
        dispatch({
            type: ActionTypes.RECEIVED_FOCUSED_POST,
            data: postId,
            channelId,
        });

        dispatch(loadChannelsForCurrentUser());
        dispatch(getChannelStats(channelId));
    };
}
