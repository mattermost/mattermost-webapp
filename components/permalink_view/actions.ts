// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {getChannel, getChannelMember, selectChannel, joinChannel, getChannelStats} from 'mattermost-redux/actions/channels';
import {getPostThread} from 'mattermost-redux/actions/posts';
import {getMissingProfilesByIds} from 'mattermost-redux/actions/users';
import {getCurrentTeam, getTeam} from 'mattermost-redux/selectors/entities/teams';
import {getCurrentUser} from 'mattermost-redux/selectors/entities/users';
import {getCurrentChannel} from 'mattermost-redux/selectors/entities/channels';
import {isCollapsedThreadsEnabled} from 'mattermost-redux/selectors/entities/preferences';
import {getUserIdFromChannelName} from 'mattermost-redux/utils/channel_utils';
import {DispatchFunc, GetStateFunc} from 'mattermost-redux/types/actions';

import {Post} from '@mattermost/types/posts';
import {Channel} from '@mattermost/types/channels';

import {loadChannelsForCurrentUser} from 'actions/channel_actions';
import {loadNewDMIfNeeded, loadNewGMIfNeeded} from 'actions/user_actions';
import {selectPostAndHighlight} from 'actions/views/rhs';

import {getHistory} from 'utils/browser_history';
import {joinPrivateChannelPrompt} from 'utils/channel_utils';
import {ActionTypes, Constants, ErrorPageTypes} from 'utils/constants';
import {isSystemAdmin} from 'mattermost-redux/utils/user_utils';
import {isComment, getPostURL} from 'utils/post_utils';
import {GlobalState} from 'types/store';

let privateChannelJoinPromptVisible = false;

type Option = {
    skipRedirectReplyPermalink: boolean;
}

function focusRootPost(post: Post, channel: Channel) {
    return async (dispatch: DispatchFunc, getState: GetStateFunc) => {
        const postURL = getPostURL(getState() as GlobalState, post);

        dispatch(selectChannel(channel.id));
        dispatch({
            type: ActionTypes.RECEIVED_FOCUSED_POST,
            data: post.id,
            channelId: channel.id,
        });

        getHistory().replace(postURL);
        return {data: true};
    };
}

function focusReplyPost(post: Post, channel: Channel, teamId: string, returnTo: string, option: Option) {
    return async (dispatch: DispatchFunc, getState: GetStateFunc) => {
        const {data} = await dispatch(getPostThread(post.root_id));

        if (data.first_inaccessible_post_time) {
            getHistory().replace(`/error?type=${ErrorPageTypes.CLOUD_ARCHIVED}&returnTo=${returnTo}`);
            return {data: false};
        }

        const state = getState() as GlobalState;

        const team = getTeam(state, channel.team_id || teamId);
        const currentChannel = getCurrentChannel(state);

        const sameTeam = currentChannel && currentChannel.team_id === team.id;

        const {skipRedirectReplyPermalink} = option;

        if (!sameTeam) {
            dispatch(selectChannel(channel.id));
        }

        if (sameTeam && returnTo && !skipRedirectReplyPermalink) {
            getHistory().replace(returnTo);
        } else if (!sameTeam || !skipRedirectReplyPermalink) {
            const postURL = getPostURL(state, post);
            getHistory().replace(postURL);
        }

        dispatch(selectPostAndHighlight(post));
        return {data: true};
    };
}

export function focusPost(postId: string, returnTo = '', currentUserId: string, option: Option = {skipRedirectReplyPermalink: false}) {
    return async (dispatch: DispatchFunc, getState: GetStateFunc) => {
        // Ignore if prompt is still visible
        if (privateChannelJoinPromptVisible) {
            return;
        }
        const {data} = await dispatch(getPostThread(postId));

        if (!data) {
            getHistory().replace(`/error?type=${ErrorPageTypes.PERMALINK_NOT_FOUND}&returnTo=${returnTo}`);
            return;
        }

        if (data.first_inaccessible_post_time) {
            getHistory().replace(`/error?type=${ErrorPageTypes.CLOUD_ARCHIVED}&returnTo=${returnTo}`);
            return;
        }

        const state = getState();
        const isCollapsed = isCollapsedThreadsEnabled(state);

        const channelId = data.posts[data.order[0]].channel_id;
        let channel = state.entities.channels.channels[channelId];
        const currentTeam = getCurrentTeam(state);
        const teamId = currentTeam.id;

        if (!channel) {
            const {data: channelData} = await dispatch(getChannel(channelId));

            if (!channelData) {
                getHistory().replace(`/error?type=${ErrorPageTypes.PERMALINK_NOT_FOUND}&returnTo=${returnTo}`);
                return;
            }

            channel = channelData;
        }

        let myMember = state.entities.channels.myMembers[channelId];

        if (!myMember) {
            // If it's a DM or GM channel and we don't have a channel member for it already, user is not a member
            if (channel.type === Constants.DM_CHANNEL || channel.type === Constants.GM_CHANNEL) {
                getHistory().replace(`/error?type=${ErrorPageTypes.PERMALINK_NOT_FOUND}&returnTo=${returnTo}`);
                return;
            }

            const membership = await dispatch(getChannelMember(channel.id, currentUserId));
            if ('data' in membership) {
                myMember = membership.data;
            }

            if (!myMember) {
                // Prompt system admin before joining the private channel
                const user = getCurrentUser(state);
                if (channel.type === Constants.PRIVATE_CHANNEL && isSystemAdmin(user.roles)) {
                    privateChannelJoinPromptVisible = true;
                    const joinPromptResult = await dispatch(joinPrivateChannelPrompt(currentTeam, channel));
                    privateChannelJoinPromptVisible = false;
                    if ('data' in joinPromptResult && !joinPromptResult.data.join) {
                        return;
                    }
                }
                await dispatch(joinChannel(currentUserId, '', channelId, channel.name));
            }
        }

        if (channel.team_id && channel.team_id !== teamId) {
            getHistory().replace(`/error?type=${ErrorPageTypes.PERMALINK_NOT_FOUND}&returnTo=${returnTo}`);
            return;
        }

        if (channel && channel.type === Constants.DM_CHANNEL) {
            const userId = getUserIdFromChannelName(currentUserId, channel.name);
            await dispatch(getMissingProfilesByIds([userId]));
            dispatch(loadNewDMIfNeeded(channel.id));
        } else if (channel && channel.type === Constants.GM_CHANNEL) {
            dispatch(loadNewGMIfNeeded(channel.id));
        }

        const post = data.posts[postId];

        if (isCollapsed && isComment(post)) {
            const {data} = await dispatch(focusReplyPost(post, channel, teamId, returnTo, option));
            if (!data) {
                return;
            }
        } else {
            dispatch(focusRootPost(post, channel));
        }

        dispatch(loadChannelsForCurrentUser());
        dispatch(getChannelStats(channelId));
    };
}
