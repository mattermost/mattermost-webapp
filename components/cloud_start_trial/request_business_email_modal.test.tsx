// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';

import {Provider} from 'react-redux';
import {act} from 'react-dom/test-utils';

import {shallow} from 'enzyme';

import * as cloudActions from 'actions/cloud';

import GenericModal from 'components/generic_modal';

import {mountWithIntl} from 'tests/helpers/intl-test-helper';
import mockStore from 'tests/test_store';

import RequestBusinessEmailModal from './request_business_email_modal';

jest.useFakeTimers();
jest.mock('lodash/debounce', () => jest.fn((fn) => fn));

describe('components/request_business_email_modal/request_business_email_modal', () => {
    const state = {
        entities: {
            users: {
                currentUserId: 'current_user_id',
            },
            admin: {},
            general: {
                license: {
                    IsLicensed: 'true',
                    Cloud: 'true',
                },
                config: {},
            },
            cloud: {
                subscription: {id: 'subscriptionID'},
            },
        },
        views: {
            modals: {
                modalState: {
                    request_business_email_modal: {
                        open: 'true',
                    },
                },
            },
        },
    };

    const props = {
        onExited: jest.fn(),
    };

    const store = mockStore(state);

    test('should match snapshot', () => {
        const wrapper = shallow(
            <Provider store={store}>
                <RequestBusinessEmailModal {...props}/>
            </Provider>,
        );
        expect(wrapper).toMatchSnapshot();
    });

    test('should show the Start Cloud Trial Button', async () => {
        const wrapper = mountWithIntl(
            <Provider store={store}>
                <RequestBusinessEmailModal {...props}/>
            </Provider>,
        );

        await act(async () => {
            const startTrialBtn = wrapper.find('CloudStartTrialButton');
            expect(startTrialBtn).toHaveLength(1);
        });
    });

    test('should call on close', async () => {
        const mockOnClose = jest.fn();

        const wrapper = mountWithIntl(
            <Provider store={store}>
                <RequestBusinessEmailModal
                    {...props}
                    onClose={mockOnClose}
                />
            </Provider>,
        );

        await act(async () => {
            wrapper.find(GenericModal).props().onExited();
            expect(mockOnClose).toHaveBeenCalled();
        });
    });

    test('should call on exited', async () => {
        const mockOnExited = jest.fn();

        const wrapper = mountWithIntl(
            <Provider store={store}>
                <RequestBusinessEmailModal
                    {...props}
                    onExited={mockOnExited}
                />
            </Provider>,
        );

        await act(async () => {
            wrapper.find(GenericModal).props().onExited();
            expect(mockOnExited).toHaveBeenCalled();
        });
    });

    test('should show the Input to enter the valid Business Email', async () => {
        const wrapper = mountWithIntl(
            <Provider store={store}>
                <RequestBusinessEmailModal {...props}/>
            </Provider>,
        );

        await act(async () => {
            expect(wrapper.find('InputBusinessEmail')).toHaveLength(1);
        });
    });

    test('should start with Start Cloud Trial Button disabled', async () => {
        const wrapper = mountWithIntl(
            <Provider store={store}>
                <RequestBusinessEmailModal {...props}/>
            </Provider>,
        );

        await act(async () => {
            const startTrialBtn = wrapper.find('CloudStartTrialButton');
            expect(startTrialBtn.props().disabled).toEqual(true);
        });
    });

    test('should ENABLE the trial button if email is VALID', async () => {
        // mock validation response to TRUE meaning the email is a valid email
        const validateBusinessEmail = () => () => Promise.resolve(true);
        jest.spyOn(cloudActions, 'validateBusinessEmail').mockImplementation(validateBusinessEmail);

        const event = {
            target: {value: 'valid-email@domain.com'},
        };

        const wrapper = mountWithIntl(
            <Provider store={store}>
                <RequestBusinessEmailModal {...props}/>
            </Provider>,
        );

        await act(async () => {
            const inputBusinessEmail = wrapper.find('InputBusinessEmail');
            const input = inputBusinessEmail.find('input');
            input.find('input').at(0).simulate('change', event);
        });

        act(() => {
            wrapper.update();
            const startTrialBtn = wrapper.find('CloudStartTrialButton');
            expect(startTrialBtn.props().disabled).toEqual(false);
        });
    });

    test('should show the success custom message if the email is valid', async () => {
        // mock validation response to TRUE meaning the email is a valid email
        const validateBusinessEmail = () => () => Promise.resolve(true);

        jest.spyOn(cloudActions, 'validateBusinessEmail').mockImplementation(validateBusinessEmail);

        const event = {
            target: {value: 'valid-email@domain.com'},
        };

        const wrapper = mountWithIntl(
            <Provider store={store}>
                <RequestBusinessEmailModal {...props}/>
            </Provider>,
        );

        await act(async () => {
            const inputBusinessEmail = wrapper.find('InputBusinessEmail');
            const input = inputBusinessEmail.find('input');
            input.find('input').at(0).simulate('change', event);
        });

        act(() => {
            wrapper.update();
            const customMessageElement = wrapper.find('.Input___customMessage.Input___success');
            expect(customMessageElement.length).toBe(1);
        });
    });

    test('should DISABLE the trial button if email is INVALID', async () => {
        // mock validation response to FALSE meaning the email is an invalid email
        const validateBusinessEmail = () => () => Promise.resolve(false);
        jest.spyOn(cloudActions, 'validateBusinessEmail').mockImplementation(validateBusinessEmail);

        const event = {
            target: {value: 'INvalid-email@domain.com'},
        };

        const wrapper = mountWithIntl(
            <Provider store={store}>
                <RequestBusinessEmailModal {...props}/>
            </Provider>,
        );

        await act(async () => {
            const inputBusinessEmail = wrapper.find('InputBusinessEmail');
            const input = inputBusinessEmail.find('input');
            input.find('input').at(0).simulate('change', event);
        });

        act(() => {
            wrapper.update();
            const startTrialBtn = wrapper.find('CloudStartTrialButton');
            expect(startTrialBtn.props().disabled).toEqual(true);
        });
    });

    test('should show the error custom message if the email is invalid', async () => {
        // mock validation response to FALSE meaning the email is an invalid email
        const validateBusinessEmail = () => () => Promise.resolve(false);
        jest.spyOn(cloudActions, 'validateBusinessEmail').mockImplementation(validateBusinessEmail);

        const event = {
            target: {value: 'INvalid-email@domain.com'},
        };

        const wrapper = mountWithIntl(
            <Provider store={store}>
                <RequestBusinessEmailModal {...props}/>
            </Provider>,
        );

        await act(async () => {
            const inputBusinessEmail = wrapper.find('InputBusinessEmail');
            const input = inputBusinessEmail.find('input');
            input.find('input').at(0).simulate('change', event);
        });

        act(() => {
            wrapper.update();
            const customMessageElement = wrapper.find('.Input___customMessage.Input___error');
            expect(customMessageElement.length).toBe(1);
        });
    });
});
