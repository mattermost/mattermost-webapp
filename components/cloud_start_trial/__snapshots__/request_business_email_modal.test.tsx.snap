// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/request_business_email_modal/request_business_email_modal should match snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <RequestBusinessEmailModal
    onExited={[MockFunction]}
  />
</ContextProvider>
`;
